api/authenticator.py

```diff
@@ -1949,11 +1887,45 @@ def scrub_credential(self, value):
     def authenticate(self, _db, credentials):
         """Turn a set of credentials into a Patron object.
 
-        :param credentials: A dictionary with keys `username` and `password`.
+        :param credentials:
+            A dictionary with keys `username` and `password`
+            or a bearer token string.
 
         :return: A Patron if one can be authenticated; a ProblemDetail
             if an error occurs; None if the credentials are missing or wrong.
         """
+        if isinstance(credentials, str):
+            return self._authenticate_from_token(_db, credentials)
+
+        elif isinstance(credentials, dict):
+            return self._authenticate_from_credentials(_db, credentials)
+
+    def _authenticate_from_token(self, _db, credentials):
+        """Turn a bearer token into a Patron object.
+
+        :param credentials: A bearer token string
+
+        :return: A Patron if one can be looked up; a ProblemDetail
+            if an error occurs.
+        """
+        credential = Credential.lookup_by_token(
+            _db, None, BasicAuthenticationProvider.TOKEN_TYPE, credentials
+        )
+
+        if isinstance(credential, Credential):
+            return credential.patron
+        else:
+            return EXPIRED_HTTP_BASIC_BEARER_TOKEN
```

Comment:
Do you need to distinguish between the case where the token has expired and the case where it's just invalid? If we're only handling one case I think it's better to just say that the token is invalid.